packages/grafana-data/src/field/displayProcessor.ts

```diff
@@ -31,80 +27,81 @@ const timeFormats: KeyValue<boolean> = {
-      } else if (field.unit && field.unit.startsWith('time:')) {
-        // Also OK
-      } else {
-        field.unit = `time:${DEFAULT_DATE_TIME_FORMAT}`;
-      }
+  if (!options || _.isEmpty(options) || !options.field) {
+    return toStringProcessor;
+  }
+  const { field } = options;
+  const config = field.config ?? {};
+
+  if (field.type === FieldType.time) {
+    if (config.unit && timeFormats[config.unit]) {
+      // Currently selected unit is valid for time fields
+    } else if (config.unit && config.unit.startsWith('time:')) {
+      // Also OK
+    } else {
+      config.unit = `time:${DEFAULT_DATE_TIME_FORMAT}`;
     }
+  }
 
-    const formatFunc = getValueFormat(field.unit || 'none');
-
-    return (value: any) => {
-      const { theme } = options;
-      const { mappings, thresholds } = field;
-      let color;
+  const formatFunc = getValueFormat(config.unit || 'none');
+  const scaleFunc = getScaleCalculator(field as Field, options.theme);
 
-      let text = _.toString(value);
-      let numeric = toNumber(value);
-      let prefix: string | undefined = undefined;
-      let suffix: string | undefined = undefined;
+  return (value: any) => {
+    const { mappings } = config;
 
-      let shouldFormat = true;
-      if (mappings && mappings.length > 0) {
-        const mappedValue = getMappedValue(mappings, value);
+    let text = _.toString(value);
+    let numeric = toNumber(value);
+    let prefix: string | undefined = undefined;
+    let suffix: string | undefined = undefined;
 
-        if (mappedValue) {
-          text = mappedValue.text;
-          const v = toNumber(text);
+    let shouldFormat = true;
+    if (mappings && mappings.length > 0) {
+      const mappedValue = getMappedValue(mappings, value);
 
-          if (!isNaN(v)) {
-            numeric = v;
-          }
+      if (mappedValue) {
+        text = mappedValue.text;
+        const v = toNumber(text);
 
-          shouldFormat = false;
+        if (!isNaN(v)) {
+          numeric = v;
         }
+
+        shouldFormat = false;
       }
+    }
 
-      if (!isNaN(numeric)) {
-        if (shouldFormat && !_.isBoolean(value)) {
-          const { decimals, scaledDecimals } = getDecimalsForValue(value, field.decimals);
-          const v = formatFunc(numeric, decimals, scaledDecimals, options.isUtc);
-          text = v.text;
-          suffix = v.suffix;
-          prefix = v.prefix;
-
-          // Check if the formatted text mapped to a different value
-          if (mappings && mappings.length > 0) {
-            const mappedValue = getMappedValue(mappings, text);
-            if (mappedValue) {
-              text = mappedValue.text;
-            }
+    if (!isNaN(numeric)) {
+      if (shouldFormat && !_.isBoolean(value)) {
+        const { decimals, scaledDecimals } = getDecimalsForValue(value, config.decimals);
+        const v = formatFunc(numeric, decimals, scaledDecimals, options.isUtc);
```

Comment:
this reminds me we forgot to change the isUtc to timezone string in the valueFormats func :)

Reply:
I am intentionally waiting till after this is merged :)